prog/dftb+/lib_dftbplus/parser.F90

```diff
@@ -6251,4 +6255,151 @@ subroutine parseRangeSeparated(node, input)
 
   end subroutine parseRangeSeparated
 
+
+  !> Reads the REKS block
+  subroutine readREKS(node, ctrl, geo)
+
+    !> Node to parse
+    type(fnode), pointer, intent(in) :: node
+
+    !> Control structure to fill
+    type(TControl), intent(inout) :: ctrl
+
+    !> geometry of the system
+    type(TGeometry), intent(in) :: geo
+
+    type(fnode), pointer :: child22, child44
+
+    ! SSR(2,2) or SSR(4,4) stuff
+    call getChild(node, "SSR22", child22, requested=.false.)
+    call getChild(node, "SSR44", child44, requested=.false.)
+
+    if (associated(child22)) then
+
+      ctrl%reksIni%tREKS = .true.
+      ctrl%reksIni%tSSR22 = .true.
+      call readSSR22(child22, ctrl, geo)
+
+    else if (associated(child44)) then
+
+      ctrl%reksIni%tREKS = .true.
+      ctrl%reksIni%tSSR44 = .true.
+      call detailedError(child44, "SSR(4,4) is not implemented yet.")
+
+    end if
+
+  end subroutine readREKS
+
+
+  !> Reads the SSR(2,2) block
+  subroutine readSSR22(node, ctrl, geo)
+
+    !> Node to parse
+    type(fnode), pointer, intent(in) :: node
+
+    !> Control structure to fill
+    type(TControl), intent(inout) :: ctrl
+
+    !> geometry of the system
+    type(TGeometry), intent(in) :: geo
+
+    !> Minimized energy functional; 1: PPS, 2: (PPS+OSS)/2
+    call getChildValue(node, "EnergyFunctional", ctrl%reksIni%Efunction, default=2)
+    !> Calculated energy states in SA-REKS
+    !> 1: calculate states included in 'EnergyFuncitonal'
+    !> 2: calculate all possible states
+    call getChildValue(node, "EnergyLevel", ctrl%reksIni%Elevel, default=1)
+    !> Calculate SSR state (SI term is included)
+    !> 1: calculate SSR state, 0: calculate SA-REKS state
+    call getChildValue(node, "UseSSRstate", ctrl%reksIni%useSSR, default=1)
+
+    !> Target SSR state
+    call getChildValue(node, "TargetState", ctrl%reksIni%rstate, default=1)
+    !> Target microstate
+    call getChildValue(node, "TargetStateL", ctrl%reksIni%Lstate, default=0)
```

Comment:
@aradi Here, L index indicate the micorstate. In SSR(2,2), there are 6 microstaes. First microstate is perfectly spin-paired configuration, second one is doubly excited configuration, third and fourth are open shell configuration, and finally fifth and sixth are triplet configuration.

So, when we want to calculate the gradient (or property) for certain microstate such as triplet configuration, then `TargetStateL` should be 5.

For overall coding and documentation, I followed the paper (JCTC, 2019, 15, 3021) for index. (e.g. T derivative, index L, etc.)

Reply:
OK, but why not using `TargetSsrState` and `TargetMicroState` as tags? Then it would be more descriptive, wouldn't it? (And you could even spare the two comment lines in the code :wink: )

Reply:
@aradi I will follow your suggestion if the exact form for `Energy` and `Gradient` input block is determined. For the inputs, I will modify all changes at one time. (as well for test files)